xbmc/cores/VideoPlayer/VideoRenderers/VideoShaders/ConversionMatrix.h

```diff
@@ -8,66 +8,102 @@
 #include <memory>
 
 extern "C" {
 #include <libavutil/pixfmt.h>
 }
 
-template <unsigned Order>
+template<uint8_t Order>
 class CMatrix
 {
 public:
-  CMatrix(float (&src)[Order][Order]);
-  CMatrix(float (&src)[Order-1][Order-1]);
-  CMatrix& operator=(const CMatrix& src);
-  CMatrix& operator=(const float (&src)[Order-1][Order-1]);
+  CMatrix() = default;
+  explicit CMatrix(const CMatrix<Order - 1>& other);
+  explicit CMatrix(const std::array<std::array<float, Order>, Order>& other) { m_mat = other; }
+  explicit CMatrix(const std::array<std::array<float, Order - 1>, Order - 1>& other);
   virtual ~CMatrix() = default;
 
-  float (&Get())[Order][Order];
-  CMatrix Invert();
+  virtual CMatrix operator*(const std::array<std::array<float, Order>, Order>& other);
+
   CMatrix operator*(const CMatrix& other);
-  CMatrix operator*=(const CMatrix& other);
-  virtual CMatrix operator*(const float (&other)[Order][Order]);
+  CMatrix& operator*=(const CMatrix& other);
 
-protected:
-  CMatrix() = default;
+  CMatrix& operator=(const std::array<std::array<float, Order - 1>, Order - 1>& other);
+
+  bool operator==(const CMatrix<Order>& other) const
+  {
+    for (int i = 0; i < Order; ++i)
+    {
+      for (int j = 0; j < Order; ++j)
+      {
+        if (m_mat[i][j] == other.m_mat[i][j])
+          continue;
+
+        // some floating point comparisions should be done by checking if the difference is within a tolerance
+        if (std::abs(m_mat[i][j] - other.m_mat[i][j]) <=
+            (std::max(std::abs(other.m_mat[i][j]), std::abs(m_mat[i][j])) * 1e-2))
+          continue;
+
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  std::array<float, Order>& operator[](int index) { return m_mat[index]; }
+
+  const std::array<float, Order>& operator[](int index) const { return m_mat[index]; }
+
+  std::array<std::array<float, Order>, Order>& Get();
 
-  void Copy(float (&dst)[Order][Order], const float (&src)[Order][Order]);
-  void Invert(float (&dst)[Order][Order], float (&src)[Order][Order]);
+  const std::array<std::array<float, Order>, Order>& Get() const;
 
-  float m_mat[Order][Order] = {};
+  CMatrix& Invert() const;
+
+  float* ToRaw() { return &m_mat[0][0]; }
+
+protected:
+  std::array<std::array<float, Order>, Order> Invert(
+      std::array<std::array<float, Order>, Order>& other) const;
+
+  std::array<std::array<float, Order>, Order> m_mat{{}};
 };
 
 class CGlMatrix : public CMatrix<4>
 {
 public:
   CGlMatrix() = default;
-  CGlMatrix(float (&src)[3][3]);
+  explicit CGlMatrix(const CMatrix<3>& other);
+  explicit CGlMatrix(const std::array<std::array<float, 3>, 3>& other);
   ~CGlMatrix() override = default;
-  CMatrix operator*(const float (&other)[4][4]) override;
+  CMatrix operator*(const std::array<std::array<float, 4>, 4>& other) override;
 };
 
 class CScale : public CGlMatrix
 {
 public:
-  CScale(float x, float y, float z);
+  explicit CScale(float x, float y, float z);
+  CScale() = delete;
```

Comment:
is this necessary? shouldn't defining a constructor with arguments automatically delete the no-argument default constructor?

Reply:
I'm not actually sure

Reply:
https://en.cppreference.com/w/cpp/language/default_constructor

"**If no user-declared constructors of any kind are provided** for a class type (struct, class, or union), the compiler will always declare a default constructor as an inline public member of its class." (emphasis added)

Reply:
is there any harm though?

Reply:
as much harm as in any other piece of dead code I guess?

Reply:
okie